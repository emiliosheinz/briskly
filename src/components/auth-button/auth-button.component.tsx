import { signIn, signOut, useSession } from 'next-auth/react'

export function AuthButton() {
  const { data: session } = useSession()

  const onClick = session ? () => signOut() : () => signIn()
  const label = session ? 'Sair' : 'Login'
  const sessionBasedClassNames = session
    ? 'text-error-700 border-error-700'
    : 'text-primary-900 border-primary-900'

  return (
    <button
      className={`bg-transparent w-full rounded border py-2 px-4 font-semibold ${sessionBasedClassNames}`}
      data-testid='auth-button'
      onClick={onClick}
    >
      {label}
    </button>
  )
}
