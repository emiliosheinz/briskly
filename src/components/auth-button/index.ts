export { AuthButton } from './auth-button.component'
