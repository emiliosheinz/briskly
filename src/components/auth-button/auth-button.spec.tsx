import { fireEvent, render } from '@testing-library/react'
import { signIn, signOut } from 'next-auth/react'

import { withNextAuth } from '~/utils/tests/with-next-auth'

import { AuthButton } from './auth-button.component'

jest.mock('next-auth/react', () => {
  const originalModule = jest.requireActual('next-auth/react')

  return {
    ...originalModule,
    signIn: jest.fn(),
    signOut: jest.fn(),
  }
})

describe('Auth Button Component', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should be properly rendered', () => {
    const component = render(withNextAuth(<AuthButton />, 'authenticated'))

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should render Sair when user is authenticated', () => {
    const component = render(withNextAuth(<AuthButton />, 'authenticated'))

    expect(component.getByText('Sair')).toBeInTheDocument()
  })

  it('should render Login when user is unauthenticated', () => {
    const component = render(withNextAuth(<AuthButton />, 'unauthenticated'))

    expect(component.getByText('Login')).toBeInTheDocument()
  })

  it('should render Login when auth status is Loading', () => {
    const component = render(withNextAuth(<AuthButton />, 'loading'))

    expect(component.getByText('Login')).toBeInTheDocument()
  })

  it('should call signOut when user is authenticated and button gets pressed', () => {
    const component = render(withNextAuth(<AuthButton />, 'authenticated'))

    const button = component.getByTestId('auth-button')
    fireEvent.click(button)

    expect(signOut).toHaveBeenCalledTimes(1)
    expect(signIn).toHaveBeenCalledTimes(0)
  })

  it('should call signIn when user is unauthenticated and button gets pressed', () => {
    const component = render(withNextAuth(<AuthButton />, 'unauthenticated'))

    const button = component.getByTestId('auth-button')
    fireEvent.click(button)

    expect(signOut).toHaveBeenCalledTimes(0)
    expect(signIn).toHaveBeenCalledTimes(1)
  })
})
