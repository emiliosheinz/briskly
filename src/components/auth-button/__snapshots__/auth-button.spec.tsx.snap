// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Button Component should be properly rendered 1`] = `
<DocumentFragment>
  <button
    class="flex flex-row gap-2 rounded-md px-6 py-2 text-base font-normal capitalize tracking-wide ring-2 transition-colors duration-150 ease-in-out hover:shadow-md bg-primary-50 hover:bg-error-50 text-error-700 ring-error-700 w-full justify-center"
    data-testid="auth-button"
  >
    Sair
  </button>
</DocumentFragment>
`;
